pkg/history/controller_history.go

```diff
@@ -0,0 +1,209 @@
+	if hs, found := rhs.Labels[ControllerRevisionHashLabel]; found {
+		hash, err := strconv.ParseInt(hs, 10, 32)
+		if err == nil {
+			rhsHash = new(uint32)
+			*rhsHash = uint32(hash)
+		}
+	}
+	if lhsHash != nil && rhsHash != nil && *lhsHash != *rhsHash {
+		return false
+	}
+	return bytes.Equal(lhs.Data.Raw, rhs.Data.Raw) && apiequality.Semantic.DeepEqual(lhs.Data.Object, rhs.Data.Object)
+}
+
+type realHistory struct {
+	client.Client
+	context context.Context
+}
+
+// NewHistory returns an instance of Interface that uses client to communicate with the API Server and lister to list
+// ControllerRevisions. This method should be used to create an Interface for all scenarios other than testing.
+func NewHistory(context context.Context, k8sclient client.Client) *realHistory {
+	return &realHistory{k8sclient, context}
+}
+
+// ListControllerRevisions lists all ControllerRevisions matching selector and owned by parent or no other
+// controller. If the returned error is nil the returned slice of ControllerRevisions is valid. If the
+// returned error is not nil, the returned slice is not valid.
+func (rh *realHistory) ListControllerRevisions(parent metav1.Object, selector labels.Selector) ([]*appsv1.ControllerRevision, error) {
+	// List all revisions in the namespace that match the selector
+	revisionList := new(appsv1.ControllerRevisionList)
+	err := rh.List(rh.context, revisionList, client.InNamespace(parent.GetNamespace()), client.MatchingLabelsSelector{Selector: selector})
+	if err != nil {
+		return nil, err
+	}
+	history := revisionList.Items
+	var owned []*appsv1.ControllerRevision
+	for i := range history {
+		ref := metav1.GetControllerOfNoCopy(&history[i])
+		if ref == nil || ref.UID == parent.GetUID() {
+			owned = append(owned, &history[i])
+		}
+
+	}
+	return owned, err
+}
+
+// CreateControllerRevision attempts to create the revision as owned by parent via a ControllerRef. Implementations may
+// cease to attempt to retry creation after some number of attempts and return an error. If the returned
+// error is not nil, creation failed. If the returned error is nil, the returned ControllerRevision has been
+// created.
+func (rh *realHistory) CreateControllerRevision(parent metav1.Object, revision *appsv1.ControllerRevision) (*appsv1.ControllerRevision, error) {
+	ns := parent.GetNamespace()
+	err := rh.Create(rh.context, revision)
+	if errors.IsAlreadyExists(err) {
+		exists := &appsv1.ControllerRevision{}
+		err := rh.Get(rh.context, types.NamespacedName{Namespace: ns, Name: revision.Name}, exists)
+		if err != nil {
+			return nil, err
+		}
+		if bytes.Equal(exists.Data.Raw, revision.Data.Raw) {
+			return exists, nil
+		} else {
+			// Since the contents of the revision are used to create the hash, the only way this
+			// happens is if the contents of the revision were changed, which is unintended behavior
+			return nil, fmt.Errorf("controller Revision with same name but different content exists")
+		}
+	}
+	return revision, err
```

Comment:
In the case where the webhook defaults, the revision itself would be different, since we store the NetworkConfig in `data.Raw` which is what we use to hash the revision,, so `errors.IsAlreadyexists(err)` returns false. There is no case where there is a match though right? Since there should be a 1:1 mapping?